File: scripts/compare_poseidon_direct.zig
const std = @import("std");
const hash_zig = @import("hash-zig");
const HashTreeOpening = hash_zig.signature.HashTreeOpening;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len < 5) {
        std.debug.print("Usage: {s} <pk_json> <sig_bin> <message> <epoch> <lifetime>\n", .{args[0]});
        std.debug.print("Example: {s} /tmp/rust_public_2pow8.key.json /tmp/rust_signature_2pow8.bin \"Cross-language benchmark message\" 0 2^8\n", .{args[0]});
        std.process.exit(1);
    }

    const pk_path = args[1];
    const sig_path = args[2];
    const message = args[3];
    const epoch = try std.fmt.parseInt(u32, args[4], 10);
    const lifetime_tag = args[5];

    // Parse lifetime - use KeyLifetimeRustCompat enum (matches remote_hash_tool.zig)
    const lifetime = if (std.mem.eql(u8, lifetime_tag, "2^8"))
        hash_zig.KeyLifetimeRustCompat.lifetime_2_8
    else if (std.mem.eql(u8, lifetime_tag, "2^18"))
        hash_zig.KeyLifetimeRustCompat.lifetime_2_18
    else if (std.mem.eql(u8, lifetime_tag, "2^32"))
        hash_zig.KeyLifetimeRustCompat.lifetime_2_32
    else {
        std.debug.print("Unsupported lifetime: {s}\n", .{lifetime_tag});
        std.process.exit(1);
    };

    // Read public key
    const pk_file = try std.fs.cwd().openFile(pk_path, .{});
    defer pk_file.close();
    const pk_bytes = try pk_file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(pk_bytes);
    const pk = try hash_zig.serialization.deserializePublicKey(pk_bytes);

    // Read signature binary
    const sig_file = try std.fs.cwd().openFile(sig_path, .{});
    defer sig_file.close();
    const sig_bytes = try sig_file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(sig_bytes);

    // Initialize scheme
    var scheme = try hash_zig.GeneralizedXMSSSignatureScheme.init(allocator, lifetime);
    defer scheme.deinit();

    // Read signature
    const signature = try readSignatureBincode(
        sig_bytes,
        allocator,
        scheme.lifetime_params.rand_len_fe,
        scheme.lifetime_params.final_layer,
        scheme.lifetime_params.hash_len_fe,
        scheme.lifetime_params.dimension,
    );
    defer signature.deinit();

    // Get rho and parameter
    const rho = signature.getRho();
    const parameter = pk.parameter;

    // Encode message and epoch
    const message_bytes = try messageToBytes(message);
    const message_fe = try scheme.encodeMessage(scheme.lifetime_params.msg_len_fe, message_bytes);
    defer allocator.free(message_fe);
    const epoch_fe = try scheme.encodeEpoch(scheme.lifetime_params.tweak_len_fe, epoch);
    defer allocator.free(epoch_fe);

    // Build combined input: randomness + parameter + epoch + message + iteration
    const RAND_LEN = scheme.lifetime_params.rand_len_fe;
    const PARAMETER_LEN = scheme.lifetime_params.parameter_len;
    const TWEAK_LEN_FE = scheme.lifetime_params.tweak_len_fe;
    const MSG_LEN_FE = scheme.lifetime_params.msg_len_fe;
    const ITER_INPUT_LEN = RAND_LEN + PARAMETER_LEN + TWEAK_LEN_FE + MSG_LEN_FE + 1;

    var combined_input = try allocator.alloc(hash_zig.FieldElement, ITER_INPUT_LEN);
    defer allocator.free(combined_input);

    var input_idx: usize = 0;
    for (0..RAND_LEN) |j| {
        combined_input[input_idx] = rho[j];
        input_idx += 1;
    }
    for (0..PARAMETER_LEN) |j| {
        combined_input[input_idx] = parameter[j];
        input_idx += 1;
    }
    for (0..TWEAK_LEN_FE) |j| {
        combined_input[input_idx] = epoch_fe[j];
        input_idx += 1;
    }
    for (0..MSG_LEN_FE) |j| {
        combined_input[input_idx] = message_fe[j];
        input_idx += 1;
    }
    combined_input[input_idx] = hash_zig.FieldElement.fromCanonical(0);
    input_idx += 1;

    // Pad to 24
    var padded_input: [24]hash_zig.FieldElement = undefined;
    for (0..ITER_INPUT_LEN) |j| {
        padded_input[j] = combined_input[j];
    }
    for (ITER_INPUT_LEN..24) |j| {
        padded_input[j] = hash_zig.FieldElement.zero();
    }

    // Print input (canonical form)
    std.debug.print("ZIG_COMPARE_INPUT (canonical, 24 values):\n", .{});
    for (0..24) |i| {
        std.debug.print("0x{x:0>8} ", .{padded_input[i].toCanonical()});
        if ((i + 1) % 8 == 0) {
            std.debug.print("\nZIG_COMPARE_INPUT (canonical): ", .{});
        }
    }
    std.debug.print("\n", .{});

    // Run poseidon compress
    const pos_outputs = try scheme.poseidon2.compress(padded_input, 15);

    // Print output (canonical form)
    std.debug.print("ZIG_COMPARE_OUTPUT (canonical, 15 values):\n", .{});
    for (0..15) |i| {
        std.debug.print("0x{x:0>8} ", .{pos_outputs[i].toCanonical()});
        if ((i + 1) % 8 == 0 and i < 14) {
            std.debug.print("\nZIG_COMPARE_OUTPUT (canonical): ", .{});
        }
    }
    std.debug.print("\n", .{});
}

fn messageToBytes(msg: []const u8) ![32]u8 {
    var bytes: [32]u8 = [_]u8{0} ** 32;
    const copy_len = @min(msg.len, 32);
    @memcpy(bytes[0..copy_len], msg[0..copy_len]);
    return bytes;
}

fn readSignatureBincode(
    bytes: []const u8,
    allocator: std.mem.Allocator,
    rand_len: usize,
    max_path_len: usize,
    hash_len: usize,
    max_hashes: usize,
) !*hash_zig.GeneralizedXMSSSignature {
    var offset: usize = 0;

    // Read path_len
    if (bytes.len < offset + 8) return error.InvalidSignature;
    const path_len = std.mem.readInt(u64, bytes[offset..][0..8], .little);
    offset += 8;
    if (path_len == 0 or path_len > max_path_len) return error.InvalidSignature;

    // Read path nodes
    var path_nodes = try allocator.alloc([8]hash_zig.FieldElement, path_len);
    errdefer allocator.free(path_nodes);
    for (0..path_len) |i| {
        if (bytes.len < offset + hash_len * 4) {
            allocator.free(path_nodes);
            return error.InvalidSignature;
        }
        for (0..hash_len) |j| {
            const montgomery = std.mem.readInt(u32, bytes[offset + j * 4 ..][0..4], .little);
            path_nodes[i][j] = hash_zig.FieldElement.fromMontgomery(montgomery);
        }
        // Pad with zeros if hash_len < 8
        for (hash_len..8) |j| {
            path_nodes[i][j] = hash_zig.FieldElement.zero();
        }
        offset += hash_len * 4;
    }

    // Read rho
    if (bytes.len < offset + rand_len * 4) {
        allocator.free(path_nodes);
        return error.InvalidSignature;
    }
    var rho: [7]hash_zig.FieldElement = undefined;
    for (0..rand_len) |i| {
        const montgomery = std.mem.readInt(u32, bytes[offset + i * 4 ..][0..4], .little);
        rho[i] = hash_zig.FieldElement.fromMontgomery(montgomery);
    }
    offset += rand_len * 4;

    // Read hashes_len
    if (bytes.len < offset + 8) {
        allocator.free(path_nodes);
        return error.InvalidSignature;
    }
    const hashes_len = std.mem.readInt(u64, bytes[offset..][0..8], .little);
    offset += 8;

    if (hashes_len == 0 or hashes_len > max_hashes) {
        allocator.free(path_nodes);
        return error.InvalidSignature;
    }
    var hashes = try allocator.alloc([8]hash_zig.FieldElement, hashes_len);
    errdefer allocator.free(hashes);

    for (0..hashes_len) |i| {
        if (bytes.len < offset + hash_len * 4) {
            allocator.free(path_nodes);
            allocator.free(hashes);
            return error.InvalidSignature;
        }
        for (0..hash_len) |j| {
            const montgomery = std.mem.readInt(u32, bytes[offset + j * 4 ..][0..4], .little);
            hashes[i][j] = hash_zig.FieldElement.fromMontgomery(montgomery);
        }
        // Pad with zeros if hash_len < 8
        for (hash_len..8) |j| {
            hashes[i][j] = hash_zig.FieldElement.zero();
        }
        offset += hash_len * 4;
    }

    // Create path
    const path = HashTreeOpening.init(allocator, path_nodes) catch |err| {
        allocator.free(path_nodes);
        allocator.free(hashes);
        return err;
    };
    allocator.free(path_nodes);

    // Note: hashes array will be freed by signature when it's deinitialized
    return hash_zig.GeneralizedXMSSSignature.init(allocator, path, rho, hashes) catch |err| {
        path.deinit();
        allocator.free(hashes);
        return err;
    };
}

File: benchmark/rust_benchmark/src/bin/compare_poseidon_direct.rs
use hashsig::symmetric::tweak_hash::poseidon::poseidon_compress;
use hashsig::symmetric::message_hash::poseidon::{encode_epoch, encode_message};
use p3_field::PrimeField32;
use p3_koala_bear::{default_koalabear_poseidon2_24, KoalaBear};
use std::env;
use std::fs;
use std::io::Read;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    
    if args.len() < 4 {
        eprintln!("Usage: {} <pk_json> <sig_bin> <message> <epoch>", args[0]);
        eprintln!("Example: {} /tmp/rust_public_2pow8.key.json /tmp/rust_signature_2pow8.bin \"Cross-language benchmark message\" 0", args[0]);
        std::process::exit(1);
    }
    
    let pk_path = &args[1];
    let sig_bin_path = &args[2];
    let message = &args[3];
    let epoch: u32 = args.get(4).and_then(|s| s.parse().ok()).unwrap_or(0);
    
    // Read public key
    let pk_json_str = fs::read_to_string(pk_path)?;
    let pk_json: serde_json::Value = serde_json::from_str(&pk_json_str)?;
    let parameter_arr = pk_json.get("parameter")
        .and_then(|v| v.as_array())
        .ok_or("Missing 'parameter' array")?;
    
    // Read signature binary
    let mut sig_file = fs::File::open(sig_bin_path)?;
    let mut sig_bytes = Vec::new();
    sig_file.read_to_end(&mut sig_bytes)?;
    
    // Parse signature binary format:
    // - path_len: u64 (little endian)
    // - path nodes: path_len * hash_len * u32 (Montgomery form)
    // - rho: rand_len * u32 (Montgomery form)
    // - hashes_len: u64 (little endian)
    // - hashes: hashes_len * hash_len * u32 (Montgomery form)
    
    let mut offset = 0;
    
    // Read path_len
    if sig_bytes.len() < offset + 8 {
        return Err("Signature too short for path_len".into());
    }
    let path_len = u64::from_le_bytes([
        sig_bytes[offset], sig_bytes[offset + 1], sig_bytes[offset + 2], sig_bytes[offset + 3],
        sig_bytes[offset + 4], sig_bytes[offset + 5], sig_bytes[offset + 6], sig_bytes[offset + 7],
    ]) as usize;
    offset += 8;
    
    // Skip path nodes (hash_len = 8 for lifetime 2^8)
    let hash_len = 8;
    offset += path_len * hash_len * 4;
    
    // Read rho (rand_len = 7 for lifetime 2^8)
    let rand_len = 7;
    if sig_bytes.len() < offset + rand_len * 4 {
        return Err("Signature too short for rho".into());
    }
    
    let mut randomness: Vec<KoalaBear> = Vec::new();
    for i in 0..rand_len {
        let montgomery = u32::from_le_bytes([
            sig_bytes[offset + i * 4],
            sig_bytes[offset + i * 4 + 1],
            sig_bytes[offset + i * 4 + 2],
            sig_bytes[offset + i * 4 + 3],
        ]);
        // Convert Montgomery to canonical for KoalaBear
        let canonical = montgomery_to_canonical(montgomery);
        randomness.push(KoalaBear::from_canonical_u32(canonical));
    }
    offset += rand_len * 4;
    
    // Extract parameter
    let mut parameter: Vec<KoalaBear> = Vec::new();
    for val in parameter_arr.iter().take(5) {
        let u32_val = val.as_u64().ok_or("Invalid parameter value")? as u32;
        parameter.push(KoalaBear::from_canonical_u32(u32_val));
    }
    
    // Encode message and epoch
    let msg_bytes = message.as_bytes();
    let message_fe = encode_message::<9>(msg_bytes);
    let epoch_fe = encode_epoch::<2>(epoch);
    let iteration_index = [KoalaBear::ZERO];
    
    // Build combined input: randomness (7) + parameter (5) + epoch (2) + message (9) + iteration (1) = 24
    let combined_input: Vec<KoalaBear> = randomness
        .iter()
        .chain(parameter.iter())
        .chain(epoch_fe.iter())
        .chain(message_fe.iter())
        .chain(iteration_index.iter())
        .copied()
        .collect();
    
    // Print input for comparison
    eprintln!("RUST_COMPARE_INPUT (canonical, 24 values):");
    for (i, fe) in combined_input.iter().enumerate() {
        eprint!("0x{:08x} ", <KoalaBear as PrimeField32>::as_canonical_u32(fe));
        if (i + 1) % 8 == 0 {
            eprintln!();
            eprint!("RUST_COMPARE_INPUT (canonical): ");
        }
    }
    eprintln!();
    
    // Run poseidon_compress
    // Use Poseidon2::new() which creates the default permutation
    use p3_symmetric::poseidon2::Poseidon2;
    let perm = Poseidon2::new();
    
    // Debug: print input length
    eprintln!("RUST_COMPARE_DEBUG: Input length: {}", combined_input.len());
    eprintln!("RUST_COMPARE_DEBUG: First 3 inputs: {:?}", 
        combined_input.iter().take(3).map(|fe| <KoalaBear as PrimeField32>::as_canonical_u32(fe)).collect::<Vec<_>>());
    
    let pos_outputs = poseidon_compress::<KoalaBear, _, 24, 15>(&perm, &combined_input);
    
    // Print output
    eprintln!("RUST_COMPARE_OUTPUT (canonical, 15 values):");
    for (i, fe) in pos_outputs.iter().enumerate() {
        eprint!("0x{:08x} ", <KoalaBear as PrimeField32>::as_canonical_u32(fe));
        if (i + 1) % 8 == 0 {
            eprintln!();
            eprint!("RUST_COMPARE_OUTPUT (canonical): ");
        }
    }
    eprintln!();
    
    Ok(())
}

// Convert Montgomery form to canonical form for KoalaBear
// KoalaBear modulus: 2^31 - 2^27 + 1 = 2130706433
fn montgomery_to_canonical(montgomery: u32) -> u32 {
    const MODULUS: u64 = 2130706433;
    const R_INV: u64 = 2147483648; // 2^31 mod MODULUS
    
    let mont_u64 = montgomery as u64;
    let canonical = (mont_u64 * R_INV) % MODULUS;
    canonical as u32
}

